### PR TITLE
Титов Семён. Задача 2. Вариант 24. Метод простых итераций

### PR DESCRIPTION
**Pull request by Semyon1104**
* ### Описание последовательной задачи: 
На вход подается матрица системы линейных уравнений, преобразуется к виду x = cx + d. Далее, начиная с точки 0 высчитываются последовательные приближения значений переменных системы уравнений. Для этого нужны 2 массива с текущими значениями и предыдущими. Процесс идет до тех пор пока максимальная разница между текущими значениями переменной и прошлыми значениями не будут меньше заданного epsilon.

* ### Описание параллельной задачи: 
На вход для удобства подается матрица A и вектор B, где СЛАУ имеет вид Ax=B, (здесь мы не преобразовываем матрицу заранее, а немного меняем вычисления новых значений). Каждой процессу выделена одна строка, если процессов меньше, то каждому процессу соответствует кратное количество строк, последний процесс забирает все оставшиеся. Если процессов выделено больше, часть из них бездействует. После подсчета очередным процессом очередного приближения переменной данные массив текущих значений x обновляется у всех процессов, по окончании итерации всеми процессами, массив с текущими значениями копируется в массив с прошлыми значениями, процесс продолжается пока в главном процессе не пройдет проверка на сходимость к epsilon.

### CODE REVIEW

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+  if (world.rank() == 0) {
+    Rows = *reinterpret_cast<size_t*>(taskData->inputs[2]);
+    epsilon_ = *reinterpret_cast<double*>(taskData->inputs[3]);
+
+    auto* Matrix_input = reinterpret_cast<double*>(taskData->inputs[0]);
+    auto* Values_input = reinterpret_cast<double*>(taskData->inputs[1]);
+
+    Matrix.assign(Matrix_input, Matrix_input + Rows * Rows);
+    Values.assign(Values_input, Values_input + Rows);
+    current.assign(Rows, 0.0);
```
Comment (AndreySorokin7) [inline tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:163]: there must be a condition under which the solution of the equation converges to an exact solution from any point, in your case from point zero

**Review by AndreySorokin7 — CHANGES_REQUESTED** [review]

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp (hunk 1 above)
Comment (Semyon1104) [reply tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:183]: fixed

File: tasks/mpi/titov_s_simple_iteration/func_tests/main.cpp
Comment (Kitos87) [inline tasks/mpi/titov_s_simple_iteration/func_tests/main.cpp:1]: Please, add more functional tests for matrix of type 1x1, 2x2, and also don't forget to check the empty matrix

**Review by Kitos87 — APPROVED** [review]

**Review by AndreySorokin7 — APPROVED** [review]

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+  if (!isDiagonallyDominant()) {
+    return false;
+  }
+
+  transformMatrix();
+
+  return true;
+}
+
+bool titov_s_simple_iteration_mpi::MPISimpleIterationSequential::validation() {
```
Comment (allnes) [inline tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:73]: Diagonal dominance is needed for convergence of the method, but where is the verification of a single solution of system linear algebraic equations?

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+    for (unsigned int j = 0; j < cols_; j++) {
+      input_[i * cols_ + j] = tmp_ptr[j];
+    }
+  }
+
+  auto* epsilon_ptr = reinterpret_cast<float*>(taskData->inputs[rows_]);
+  epsilon_ = *epsilon_ptr;
+  if (!isDiagonallyDominant()) {
+    return false;
+  }
```
Comment (allnes) [inline tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:66]: its validation

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+
+  transformMatrix();
```
Comment (allnes) [inline tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:68]: its run()

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp [hunk 5]
```diff
@@ -13,6 +13,28 @@
 
 using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp:14]: please remove it

File: tasks/seq/titov_s_simple_iteration/src/ops_seq.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
+// Copyright 2024 Nesterov Alexander
+#include "seq/titov_s_simple_iteration/include/ops_seq.hpp"
+
+#include <thread>
+
+using namespace std::chrono_literals;
```
Comment (allnes) [inline tasks/seq/titov_s_simple_iteration/src/ops_seq.cpp:6]: please remove it

Comment (Semyon1104) [reply tasks/seq/titov_s_simple_iteration/src/ops_seq.cpp]: ready

File: tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp (hunk 5 above)
Comment (Semyon1104) [reply tasks/mpi/titov_s_simple_iteration/src/ops_mpi.cpp]: ready

**Review by allnes — APPROVED** [review]

**Review by aobolensk — APPROVED** [review]
